Allow commit on different collections

jersiovic opened the pull request:
PR related with this issue https://github.com/sebastienros/yessql/issues/240

jersiovic commented on src/YesSql.Core/Session.cs:
```diff
@@ -780,17 +784,19 @@ private void BatchCommands()
             {
-                foreach (var page in createDocumentCommands.PagesOf(_store.Configuration.CommandsPageSize))
+                foreach (var page in createDocumentCommands.PagesOf(_store.Configuration.CommandsPageSize)
+                    .SplitPagesBy(c=>c.CollectionName))
```
This call to SplitPagesBy is one thing I don't feel pretty well.
The poins is I saw that you pass to dapper an string with an insert in the pseudo SQL and a collection of documents to insert. Problem I found is the SQL is not the same for all the documents because some of them are for one collection and others are for others one.
I had two options: 

- To group inserts (or updates on next foreach) to make as much as possible inserts in one batch commad. But if generates me doubts cause I'm not sure it is right to change the oreder on which things are inserted on db.
- To split each page every time I find a document with a different collection than previous one. Problem is this breaks  batch in small pieces.

What do you think, is it ok like it is?
Thank you

jersiovic commented on src/YesSql.Core/Session.cs:
```diff
@@ -780,17 +784,19 @@ private void BatchCommands()
             {
-                foreach (var page in deleteDocumentCommands.PagesOf(_store.Configuration.CommandsPageSize))
+                foreach (var page in deleteDocumentCommands.PagesOf(_store.Configuration.CommandsPageSize)
+                    .SplitPagesBy(c => c.CollectionName))
```
same here

sebastienros commented:
Notes from meeting:

We need index tables for each document table (each collection) because of foreign keys

Query<Token, TokenIndex>(collection)
Query<TokenIndex>(collection)

Document
TokenIndex

Collection1_Document
Collection1_TokenIndex

store.RegisterIndexProvider<P>(collection)

{ _saved, _updated, _delete } should be a set per collection
indexProviders should be a list per collection

jersiovic commented:
I've updated this PR with the changes I made to make that any table created, updated or queried uses as prefix, the prefix and the collection name.
So no matter if you are creating a regular table, a document table or an index table that prefix will be added.
This PR now has what I'm going to use in production together with the mechanism I was using at Orchard side.
I provide it to you only just in case it could help you to spot points that will need to be changed when you start working on what we agreed on the meeting.

Appveyor show 3 errors for postgress tests .. I will install Postgress to check what happens

jersiovic commented:
Fixed on last commit also for Postgress

jersiovic commented:
Added another commit on this PR, cause as it was it didn't work with my approach of having my own Session and SchemaBuider enveloping original objects when the type or the type name are related to a different collection prefix.
Problem was that queries like this one didn't work (without using an `using Collection collection= ...) enveloping it`:
`var users = await _session.Query<User, UserByClaimIndex>(uc => uc.ClaimType == claim.Type && uc.ClaimValue == claim.Value).ListAsync();`

My Session object (called CollectionSession)I retuns a Query enveloped  in my own ColletionQuery object.
```
...
        public IQuery Query()
        {
            return new CollectionQuery(Session.Query(), _yesSQLCollectionNameService);
        }
...
```
whic also envelops ForEach calls within a collection
```
public IQuery<T> For<T>(bool filterType = true) where T : class
        {
            var collectionName = _yesSQLCollectionNameService.GetCollectionNameForType(typeof(T));
            if (collectionName == null)
            {
                return _query.For<T>(filterType);
            }
            using (new NamedCollection(collectionName))
            {
                return _query.For<T>(filterType);
            }
        }
```
However this is not enough cause when at YesSql DefaultQuery evaluates current collection when the query is created, it still doesn't know which is the type rquested in the query. So I need to delay this check to the momento For method is called in DefaultQuery.
What I did is hacky and has potential problems, that by the moment I've seen are not happening to me. So I'm going to use it temporally while you work on the final solution that allows relate entities with collections within YesSQL
